pkg/protocol/protocol.go

```diff
@@ -2,316 +2,216 @@ package protocol
+	Engines *Engines
 
-	p.shutdown()
+	// Options contains the options that were used to create the protocol.
+	Options *Options
 
-	p.TriggerStopped()
+	// EvictionState contains the eviction state of the protocol.
+	reactive.EvictionState[iotago.SlotIndex]
 
-	return p.context.Err()
+	// ReactiveModule embeds the reactive module logic of the protocol.
+	*module.ReactiveModule
 }
 
-func (p *Protocol) linkToEngine(engineInstance *engine.Engine) {
-	p.Events.Engine.LinkTo(engineInstance.Events)
-}
+// New creates a new protocol instance from the given parameters.
+func New(logger log.Logger, workers *workerpool.Group, networkEndpoint network.Endpoint, opts ...options.Option[Protocol]) *Protocol {
+	return options.Apply(&Protocol{
+		Events:         NewEvents(),
+		Workers:        workers,
+		Options:        NewDefaultOptions(),
+		ReactiveModule: module.NewReactiveModule(logger),
+		EvictionState:  reactive.NewEvictionState[iotago.SlotIndex](),
+	}, opts, func(p *Protocol) {
+		shutdownSubComponents := p.initSubcomponents(networkEndpoint)
 
-func (p *Protocol) shutdown() {
-	if p.networkProtocol != nil {
-		p.networkProtocol.Shutdown()
-	}
+		p.Initialized.OnTrigger(func() {
+			shutdown := lo.Batch(
+				p.initEviction(),
+				p.initGlobalEventsRedirection(),
+				p.initNetwork(),
 
-	p.ChainManager.Shutdown()
-	p.Workers.Shutdown()
+				shutdownSubComponents,
+			)
 
-	p.activeEngineMutex.RLock()
-	p.mainEngine.Shutdown()
-	if p.candidateEngine != nil {
-		p.candidateEngine.engine.Shutdown()
-	}
-	p.activeEngineMutex.RUnlock()
-}
+			p.Shutdown.OnTrigger(shutdown)
+		})
 
-func (p *Protocol) initEngineManager() {
-	p.EngineManager = enginemanager.New(
-		p.Workers.CreateGroup("EngineManager"),
-		p.HandleError,
-		p.optsBaseDirectory,
-		DatabaseVersion,
-		p.optsStorageOptions,
-		p.optsEngineOptions,
-		p.optsPreSolidFilterProvider,
-		p.optsPostSolidFilterProvider,
-		p.optsBlockDAGProvider,
-		p.optsBookerProvider,
-		p.optsClockProvider,
-		p.optsBlockGadgetProvider,
-		p.optsSlotGadgetProvider,
-		p.optsSybilProtectionProvider,
-		p.optsNotarizationProvider,
-		p.optsAttestationProvider,
-		p.optsLedgerProvider,
-		p.optsSchedulerProvider,
-		p.optsTipManagerProvider,
-		p.optsTipSelectionProvider,
-		p.optsRetainerProvider,
-		p.optsUpgradeOrchestratorProvider,
-		p.optsSyncManagerProvider,
-	)
+		p.Constructed.Trigger()
 
-	mainEngine, err := p.EngineManager.LoadActiveEngine(p.optsSnapshotPath)
-	if err != nil {
-		panic(fmt.Sprintf("could not load active engine: %s", err))
-	}
-	p.mainEngine = mainEngine
+		p.waitMainEngineInitialized()
+	})
 }
 
-func (p *Protocol) initChainManager() {
-	p.ChainManager = chainmanager.NewManager(p, p.HandleError, p.optsChainManagerOptions...)
-	p.Events.ChainManager.LinkTo(p.ChainManager.Events)
+// IssueBlock issues a block to the node.
+func (p *Protocol) IssueBlock(block *model.Block) error {
+	p.Network.Events.BlockReceived.Trigger(block, "self")
```

Comment:
Because we don't just want to process the block in our engine but we also want to feed it to our latestSeenBlocks and so on that listen to these kind of network events.

Ideally you want to inject things where they would usually arrive, to make sure you do exactly the same things with all blocks.